main/util/geometryutil.ts

```diff
@@ -0,0 +1,1809 @@
+    const flatnessThreshold = 4 * tol * tol;
+    let ux = 2 * c1.x - p1.x - p2.x;
+    ux *= ux;
+    let uy = 2 * c1.y - p1.y - p2.y;
+    uy *= uy;
+    return ux + uy <= flatnessThreshold;
+  },
+
+  linearize: function (p1: Point, p2: Point, c1: Point, tol: number): Point[] {
+    const finished: Point[] = [p1];
+    const todo: QuadraticBezierSegmentDef[] = [{ p1, p2, c1 }];
+    while (todo.length > 0) {
+      const segment = todo.shift()!;
+      if (this.isFlat(segment.p1, segment.p2, segment.c1, tol)) {
+        finished.push(new Point(segment.p2.x, segment.p2.y));
+      } else {
+        const divided = this.subdivide(segment.p1, segment.p2, segment.c1, 0.5);
+        todo.unshift(divided[1]);
+        todo.unshift(divided[0]);
+      }
+    }
+    return finished;
+  },
+
+  subdivide: function (
+    p1: Point,
+    p2: Point,
+    c1: Point,
+    t: number,
+  ): [QuadraticBezierSegmentDef, QuadraticBezierSegmentDef] {
+    const mid1: Point = new Point(
+      p1.x + (c1.x - p1.x) * t,
+      p1.y + (c1.y - p1.y) * t,
+    );
+    const mid2: Point = new Point(
+      c1.x + (p2.x - c1.x) * t,
+      c1.y + (p2.y - c1.y) * t,
+    );
+    const mid3: Point = new Point(
+      mid1.x + (mid2.x - mid1.x) * t,
+      mid1.y + (mid2.y - mid1.y) * t,
+    );
+    return [
+      { p1, p2: mid3, c1: mid1 },
+      { p1: mid3, p2, c1: mid2 },
+    ];
+  },
+};
+
+export const CubicBezier = {
+  isFlat: function (
+    p1: Point,
+    p2: Point,
+    c1: Point,
+    c2: Point,
+    tol: number,
+  ): boolean {
+    const flatnessThreshold = 16 * tol * tol;
+    let ux = 3 * c1.x - 2 * p1.x - p2.x;
+    ux *= ux;
+    let uy = 3 * c1.y - 2 * p1.y - p2.y;
+    uy *= uy;
+    let vx = 3 * c2.x - 2 * p2.x - p1.x;
+    vx *= vx;
+    let vy = 3 * c2.y - 2 * p2.y - p1.y;
+    vy *= vy;
+    return Math.max(ux, vx) + Math.max(uy, vy) <= flatnessThreshold;
+  },
+
+  linearize: function (
+    p1: Point,
+    p2: Point,
+    c1: Point,
+    c2: Point,
+    tol: number,
+  ): Point[] {
+    const finished: Point[] = [p1];
+    const todo: CubicBezierSegmentDef[] = [{ p1, p2, c1, c2 }];
+    while (todo.length > 0) {
+      const segment = todo.shift()!;
+      if (this.isFlat(segment.p1, segment.p2, segment.c1, segment.c2, tol)) {
+        finished.push(new Point(segment.p2.x, segment.p2.y));
+      } else {
+        const divided = this.subdivide(
+          segment.p1,
+          segment.p2,
+          segment.c1,
+          segment.c2,
+          0.5,
+        );
+        todo.unshift(divided[1]);
+        todo.unshift(divided[0]);
+      }
+    }
+    return finished;
+  },
+
+  subdivide: function (
+    p1: Point,
+    p2: Point,
+    c1: Point,
+    c2: Point,
+    t: number,
+  ): [CubicBezierSegmentDef, CubicBezierSegmentDef] {
+    const t_p1 = p1,
+      t_c1 = c1,
+      t_c2 = c2,
+      t_p2 = p2;
+    const p1c1: Point = new Point(
+      t_p1.x + (t_c1.x - t_p1.x) * t,
+      t_p1.y + (t_c1.y - t_p1.y) * t,
+    );
```

Comment:
Should define and use a `lerp` function that encapsulates the "partway from one point to another" behavior.